runtime/lua/vim/render.lua

```diff
@@ -0,0 +1,162 @@
+    (h1 (heading) @h1)
+    (h2 (heading) @h2)
+    (h3 (heading) @h3)
+    (column_heading (heading) @h4)
+  ]],
+  markdown = [[
+    (setext_heading
+      heading_content: (_) @h1
+      (setext_h1_underline))
+    (setext_heading
+      heading_content: (_) @h2
+      (setext_h2_underline))
+
+    (atx_heading
+      (atx_h1_marker)
+      heading_content: (_) @h1)
+    (atx_heading
+      (atx_h2_marker)
+      heading_content: (_) @h2)
+    (atx_heading
+      (atx_h3_marker)
+      heading_content: (_) @h3)
+    (atx_heading
+      (atx_h4_marker)
+      heading_content: (_) @h4)
+    (atx_heading
+      (atx_h5_marker)
+      heading_content: (_) @h5)
+    (atx_heading
+      (atx_h6_marker)
+      heading_content: (_) @h6)
+  ]],
+}
+
+--- Extract headings from buffer
+--- @param bufnr integer buffer to extract headings from
+--- @return table[]? headings each entry has keys `bufnr`, `lnum`, `capture`, `text`
+--- todo(clason): memoize
+local get_headings = function(bufnr)
+  local lang = vim.treesitter.language.get_lang(vim.bo[bufnr].filetype)
+  if not lang then
+    return
+  end
+  local parser = assert(vim.treesitter.get_parser(bufnr, lang, { error = false }))
+  local query = vim.treesitter.query.parse(lang, heading_queries[lang])
+  local root = parser:parse()[1]:root()
+  local headings = {}
+  for id, node, _, _ in query:iter_captures(root, bufnr) do
+    local text = vim.treesitter.get_node_text(node, bufnr)
+    local row, col = node:start()
+    --- why can't you just be normal?!
+    local skip ---@type boolean|integer
+    if lang == 'vimdoc' then
+      -- only column_headings at col 1 are headings, otherwise it's code examples
+      skip = (id == 4 and col > 0)
+        -- ignore tabular material
+        or (id == 4 and (text:find('\t') or text:find('  ')))
+        -- ignore tag-only headings
+        or (node:child_count() == 1 and node:child(0):type() == 'tag')
+    end
+    if not skip then
+      table.insert(headings, {
+        bufnr = bufnr,
+        lnum = row + 1,
+        text = text,
+        level = id,
+      })
+    end
+  end
+  return headings
+end
+
+--- Show a table of contents for the help buffer in a loclist
+function M.show_toc()
+  local bufnr = vim.api.nvim_get_current_buf()
+  local headings = get_headings(bufnr)
+  if not headings or #headings == 0 then
+    return
+  end
+  -- add indentation for nicer list formatting
+  for _, heading in pairs(headings) do
+    if heading.level > 2 then
+      heading.text = '  ' .. heading.text
+    end
+    if heading.level > 4 then
+      heading.text = '  ' .. heading.text
+    end
+  end
+  vim.fn.setloclist(0, headings, ' ')
+  vim.fn.setloclist(0, {}, 'a', { title = 'Help TOC' })
+  vim.cmd.lopen()
+end
+
+--- Jump to section
+--- @param opts table jump options
+---  - count integer direction to jump (>0 forward, <0 backward)
+---  - level integer only consider headings up to level
+--- todo(clason): support count
+function M.jump(opts)
```

Comment:
doesn't seem like `vim.render` or `vim.ui` is the place for this.

Reply:
Again, internal utility function. Would love to have a generic `jump` or `motion` module, but that is a job for somebody else.